### PR TITLE
Make cargo aware of dwp files.

### PR DESCRIPTION
**Pull request by khuey**
When using -Csplit-debuginfo=packed on Linux, rustc will produce a dwp file. Unlike the dwo files, whose paths are embedded into the binary, there's no information in the binary to help a debugger locate a dwp file. By convention, the dwp file for `<EXE>` is given the name `<EXE>.dwp` and placed next to `<EXE>`.

When cargo hardlinks the executable file rustc put in target/debug/deps into target/debug, it also needs to hardlink the dwp file along with it. Failing to do this prevents the debugger from finding the dwp file when the binary is executed from target/debug, as there's no way for the debugger to know to look in the deps subdirectory.

The split_debuginfo option is passed down into file_types to make this possible. For cargo clean manual handling is added to match the other split_debuginfo files. bin_link_for_target also passes in None because it won't care about the dwp file.

### CODE REVIEW

**Comment by khuey** [comment]
cc @davidtwco

**Comment by khuey** [comment]
CI seems broken.

**Comment by weihanglo** [comment]
Thank you! Looks like a duplicate of <https://github.com/rust-lang/cargo/pull/11384>? I would probably ping them to see if they want to proceed. There is also a discussion regarding artifact name conflict on Windows which is worth considering.

**Comment by khuey** [comment]
Thanks for the pointer to the other PR. I hadn't seen that.

The mingw issue they ran across is actually a rustc bug: rust-lang/rust#106811.  `.dwp` should be added to the filename; it should not replace the existing extension. On Linux this isn't noticed on regular binaries but it is visible if you build a dylib. I've fixed the cargo side of that in the updated commit here.

I've also added a test along the lines you suggested in the other PR. It would be good to test the dylib behavior too but I don't know how cargo handles varying rustc versions for tests. Any test would need a version of rustc with the rustc PR above merged to pass.

I think the main difference between the PRs at this point is whether the `split_debuginfo` value is passed into `file_types` (and thus whether the DWP entry is emitted only when relevant or all the time).

**Comment by khuey** [comment]
And the value of `should_replace_hyphens` has to match the value the binary itself had. Setting it to false unconditionally as in the other PR is wrong.

**Comment by ehuss** [comment]
> I think the main difference between the PRs at this point is whether the `split_debuginfo` value is passed into `file_types` (and thus whether the DWP entry is emitted only when relevant or all the time).

So far cargo has generated the `FileType` list with all potential files that rustc *might* generate, and doesn't necessarily differentiate based on anything like profile settings. It is generally safe to list extra files since cargo will ignore them if rustc didn't generate them. I can't offhand think of a particular problem of differentiating, but by adding it here for just dwp that makes the code inconsistent with other platforms.

**Comment by khuey** [comment]
Alright, done.

**Comment by khuey** [comment]
Yay the tests finally pass.

File: src/cargo/core/compiler/build_context/target_info.rs
Comment (khuey) [reply src/cargo/core/compiler/build_context/target_info.rs:467]: .so, or .exe/.dll on mingw

File: src/cargo/core/compiler/build_context/target_info.rs
Comment (khuey) [reply src/cargo/core/compiler/build_context/target_info.rs:482]: To be honest I don't know why bins behave differently from e.g. dylibs, but what's important is that we match that behavior exactly. I've left a comment to that effect.

File: src/cargo/core/compiler/build_context/target_info.rs
Comment (khuey) [reply src/cargo/core/compiler/build_context/target_info.rs]: I've left a comment about that here, and changed this to use the implicit argument.

I did *not* change the format! I added in the implementation of cargo clean since the surrounding code doesn't use implicit arguments.

**Comment by khuey** [comment]
Also even once you are satisfied with this you probably shouldn't merge it before rust-lang/rust#106811 is merged.

**Comment by Gankra** [comment]
Oh, thanks so much for picking this up and doing it right!

**Comment by khuey** [comment]
The rustc PRs this depends on have been merged.

**Review by weihanglo — APPROVED** [review]
Looks good. Thanks!

**Pull request merged**
